BTCPayServer/Services/PayjoinClient.cs

```diff
@@ -0,0 +1,203 @@
+                }
+            }
+
+            var hex = await bpuresponse.Content.ReadAsStringAsync();
+            var newPSBT = PSBT.Parse(hex, originalTx.Network);
+
+            // Checking that the PSBT of the receiver is clean
+            if (newPSBT.GlobalXPubs.Any())
+            {
+                throw new PayjoinSenderException("GlobalXPubs should not be included in the receiver's PSBT");
+            }
+
+            if (newPSBT.Outputs.Any(o => o.HDKeyPaths.Count != 0) || newPSBT.Inputs.Any(o => o.HDKeyPaths.Count != 0))
+            {
+                throw new PayjoinSenderException("Keypath information should not be included in the receiver's PSBT");
+            }
+            ////////////
+
+            newPSBT = await _explorerClientProvider.UpdatePSBT(derivationSchemeSettings, newPSBT);
+            if (newPSBT.CheckSanity() is IList<PSBTError> errors2 && errors2.Count != 0)
+            {
+                throw new PayjoinSenderException($"The PSBT of the receiver is insane ({errors2[0]})");
+            }
+
+            // We make sure we don't sign things what should not be signed
+            foreach (var finalized in newPSBT.Inputs.Where(i => i.IsFinalized()))
+            {
+                finalized.ClearForFinalize();
+            }
+            // Make sure only the only our output have any information
+            foreach (var output in newPSBT.Outputs)
+            {
+                output.HDKeyPaths.Clear();
+                foreach (var originalOutput in  originalTx.Outputs)
+                {
+                    if (output.ScriptPubKey == originalOutput.ScriptPubKey)
+                        output.UpdateFrom(originalOutput);
+                }
+            }
+
+            // Making sure that our inputs are finalized, and that some of our inputs have not been added
+            foreach (var input in newPSBT.Inputs.CoinsFor(derivationSchemeSettings.AccountDerivation,
+                signingAccount.AccountKey, signingAccount.GetRootedKeyPath()))
+            {
+                if (originalTx.Inputs.FindIndexedInput(input.PrevOut) is PSBTInput ourInput)
+                {
+                    if (input.IsFinalized())
+                        throw new PayjoinSenderException("A PSBT input from us should not be finalized");
+                }
+                else
+                {
+                    throw new PayjoinSenderException(
+                        "The payjoin receiver added some of our own inputs in the proposal");
+                }
+            }
+
+            // Making sure that the receiver's inputs are finalized
+            foreach (var input in newPSBT.Inputs)
+            {
+                if (originalTx.Inputs.FindIndexedInput(input.PrevOut) is null && !input.IsFinalized())
+                    throw new PayjoinSenderException("The payjoin receiver included a non finalized input");
+            }
+
+            // The receiver, may have removed some of our inputs, that's fine.
```

Comment:
Why would a receiver remove a sender's inputs?

Reply:
why not, what matter is that we send what we expect to send or less.

Reply:
What if the sender wanted to consolidate his utxos while doing the payment?